scripts/qa_crowbarsetup.sh

```diff
@@ -2053,43 +2054,111 @@ function setup_trusted_cert
         local cn=$controller_nodes
     fi
 
-    if [ ! -f $cn.pem ] ; then
+    if [ ! -f rootca.pem ] ; then
+        # Generate Root CA key
+        openssl genrsa -aes256 -out rootca.key -passout pass:"crowbar" 4096
         # Generate Root CA
-        openssl req -x509 -newkey rsa:2048 -keyout rootca.key -out rootca.pem \
+        openssl req -x509 -key rootca.key -out rootca.pem \
+            -passin pass:"crowbar" \
             -days 365 -nodes \
             -subj "/C=DE/ST=Bayern/L=Nuernberg/O=CI/CN=root"
-        # Generate CSR
-        subj_alt="DNS:${cn},DNS:public-${cn},DNS:public.${cn}"
+    fi
+
+    subj_alt="DNS:${cn},DNS:public-${cn},DNS:public.${cn}"
+
+    if [ "$certtype" == "server" ] && [ ! -f $cn.server.pem ] ; then
+        # Generate server CSR
         # openssl 1.1.1 supports setting the subjectAltName directly in the command line,
         # but SLES 12 SP4 only has 1.0.2 so we need to read in the config like this
-        openssl req -new -newkey rsa:2048 -keyout $cn.key -out $cn.csr \
+        openssl req -new -newkey rsa:2048 -keyout $cn.server.key -out $cn.server.csr \
             -days 365 -nodes \
             -subj "/C=DE/ST=Bayern/L=Nuernberg/O=CI/CN=${cn}" \
             -config <(sed -e "/^\[ req \]/areq_extensions = v3_req" \
                         -e "/^\[ v3_req \]/asubjectAltName=${subj_alt}" \
                         /etc/ssl/openssl.cnf )
-        # Generate Cert
-        openssl x509 -req -in $cn.csr -CA rootca.pem -CAkey rootca.key \
-            -CAcreateserial -days 365 -out $cn.pem \
+        # Generate server Cert
+        openssl x509 -req -in $cn.server.csr -CA rootca.pem -CAkey rootca.key \
+            -passin pass:"crowbar" \
+            -CAcreateserial -days 365 -out $cn.server.pem \
             -extensions v3_ca \
             -extfile <(sed -e "/^\[ v3_ca \]/asubjectAltName=${subj_alt}" \
                         /etc/ssl/openssl.cnf )
+    fi
 
+    if [ "$certtype" == "client" ] && [ ! -f $cn.client.pem ] ; then
+        # Generate CSR for client cert
+        subj_alt="DNS:${cn},DNS:public-${cn},DNS:public.${cn}"
```

Comment:
do we really need the SAN for client cert?

Reply:
Not really sure, the documentation out there concerning client certificates is really slim. I imagine you're right and we only need the IP address, hostname or domain name wildcard in the subject CN, but in the SAN field you can actually specify what the values actually are (IP addresses, DNS, etc). I'll do some more testing when I have time, but for now I'd like to stick to something that I already know works.

Reply:
There's also the issue of generating a single client certificate that can be used for all of the controller nodes in the designate-api cluster. We can't rely on admin IP addresses for that, so we need the domain name wildcard.